app/components/Connect4/lib/matches/diagonal.js

```diff
@@ -38,36 +38,29 @@ function isTopRight(grid) {
 
     found = 0;
     foundPiece = 0;
-    col = baseCol;
+    col = baseCol - 1; // Subtracting 1 to compensate for incrementing col at
+                       // the beginning of the loop
 
-    // Here we work our way up the current diagonal
+    // Here we work our way *UP* the current diagonal
     for (let row = 0; row < numRows; row++) {
+      col++;
 
       // Ensure that the given column and row are on the board
       if (col >= 0 && col < numCols && row < numRows) {
 
         let piece = grid[col][row];
 
-        // Does nothing if piece is zero
-        if (!piece) {
-          continue;
+        if(!piece) {
+          found = 0;
         }
 
-        if (!foundPiece) {
-          foundPiece = piece;
+        if (!!piece && (piece === foundPiece || !foundPiece) && (++found) === matchReq) {
```

Comment:
Breakdown of this condition:
`!!piece` : there has to be a piece to count of course
`piece === foundPiece || !foundPiece` : if this piece is like the last one, the count may be increased. But also if the previous cell was empty (i.e. `!foundPice`), you must still increase the count `found` from `0` to `1`.
`(++found) === matchReq` : the familiar win condition